### DISCUSSION TITLE
Convert From Dec To Bin

### DISCUSSION THREAD

**Discussion by sergey-shandar** [functionalscript/functionalscript#223] [Ideas]
How to divide big floats

## Unsigned BigInt Rounding

```js
/** @typedef bigint UBigInt */

/** @type {(a: UBigInt) => (b: UBigInt) => readonly[UBigInt, UBigInt]} */
const uDiv = a => b => [a / b, a % b]

/** @type {(a: UBigInt) => (b: UBigInt) => UBigInt} */
const uRoundDiv = a => b => {
    const [d, r] = uDiv(a)(b)
    const sign = cmp(r << 1)(b)
    return d + (sign === 0 ? d & 1 : sign >> 1)
}

/** @typedef {readonly[UBigInt, UBigInt, UBigInt]} UFraction */

/** @type {(a: UFraction) => (x: UBigInt) => UFraction */
const uFractionDiv = ([d0, r0, c0]) => c1 => {
  const [d1, r1] = uDiv(d0)(c1)
  return [d1, c0 * r0 + r1, c0 * c1]
}
```

## Convert From One Base To Another

Our floating point number is represented as $m*r^e$.
 
- $m$ is an integer and $m > 0$,
- $e$ is an integer,
- $r$ is an integer and $base > 1$.

Any $m*base^e$ can be converted to a fraction of two unsigned integers.

if $e >= 0$ then $\frac{mr^e}{1}$ else $\frac{m}{r ^{-e}}$

As a result, we have $a / b$, where 
- $a$ is an integer and $a > 0$,
- $b$ is an iteger and $b > 0$.

### Converting a fraction to a floating point number

Assume $a <= b$ 

Find $m$ and $e$ where $b <= r^e < br$

$$\frac{a}{b} = \frac{m + x/b}{r^e},$$

where $r^n <= m < r^{n+1}$

**Comment by Trinidadec** [comment]
```
case 1:
m = 0b100_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0010_0101n  //59 bits
decimal exp = -1

div = 0b1100_1100_1100_1100_1100_1100_1100_1100_1100_1100_1100_1100_1101_0100n //53 + 3 bits
binary exp = -1
mod = 1n
```

if 53 bits left
`0b11001100110011001100110011001100110011001100110011010.100n`
rounding up

---

```
case 2:
m = 0b100_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0000_0010_0100n //59 bits
decimal exp = -1

div = 0b1100_1100_1100_1100_1100_1100_1100_1100_1100_1100_1100_1100_1101_0100n //53 + 3 bits
binary exp = -1
mod = 0n
```

if 53 bits left
`0b11001100110011001100110011001100110011001100110011010.100n`
rounding down (to even)

**Comment by sergey-shandar** [comment]
```js
const s = "0.1"
// [1, -1]
let [dM, dE] = parse(s) 
// increase Matnisa to 53 bits.
while (dM < 2 ** 53) {
  dM *= 10
  dE -= 1;
}
//
```

**Comment by sergey-shandar** [comment]
Another presentation of `[m, e]` is `m * base ** e` .

if `e < 0` then `m / base ** -e`

In result we have `m / b`

**Task**: `m/b`, `baseX` => `[mX, eX]`, where `mX * baseX ** eX ~ m/b`, and `mX` has only limited number of digits.

**Comment by sergey-shandar** [comment]
`10 ** 19 < 2 ** 64 = 18,446,744,073,709,551,616 < 10 ** 20`

**Comment by sergey-shandar** [comment]
```js
// a / b
// Make sure `min * b <= a < max * b` by multiplying/dividing `a` by `r` and decreasing/increasing `e`, where `min` is a minimal mantissa value.
const mantissa = uRoundDiv(a)(b)
```
